src/Database/LSMTree/Internal/Snapshot/Codec.hs

```diff
@@ -559,16 +552,36 @@ instance DecodeVersioned SuppliedCredits where
 
 -- MergeType
 
-instance Encode MergeType  where
-  encode MergeMidLevel  = encodeWord 0
-  encode MergeLastLevel = encodeWord 1
-  encode MergeUnion     = encodeWord 2
+instance Encode MR.LevelMergeType  where
+  encode MR.MergeMidLevel  = encodeWord 0
+  encode MR.MergeLastLevel = encodeWord 1
 
-instance DecodeVersioned MergeType where
+instance DecodeVersioned MR.LevelMergeType where
   decodeVersioned V0 = do
       tag <- decodeWord
       case tag of
-        0 -> pure MergeMidLevel
-        1 -> pure MergeLastLevel
-        2 -> pure MergeUnion
-        _ -> fail ("[MergeType] Unexpected tag: " <> show tag)
+        0 -> pure MR.MergeMidLevel
+        1 -> pure MR.MergeLastLevel
+        _ -> fail ("[LevelMergeType] Unexpected tag: " <> show tag)
+
+-- | We start the tags for these merge types at an offset. This way, if we
+-- serialise @MR.MergeMidLevel :: MR.LevelMergeType@ as 0 and then accidentally
+-- try deserialising it as a @MR.TreeMergeType@, that will fail.
+--
+-- However, 'MR.LevelMergeType' and 'MR.TreeMergeType' are only different
+-- (overlapping) subsets of 'MR.MergeType'. In particular, 'MR.MergeLastLevel'
+-- and 'MR.MergeLevel' are semantically the same. Encoding them as the same
+-- number leaves the door open to relaxing the restrictions on which merge types
+-- can occur where, e.g. decoding them as a general 'MR.MergeType', without
+-- having to change the file format.
+instance Encode MR.TreeMergeType  where
+  encode MR.MergeLevel = encodeWord 1
+  encode MR.MergeUnion = encodeWord 2
+
+instance DecodeVersioned MR.TreeMergeType where
+  decodeVersioned V0 = do
+      tag <- decodeWord
+      case tag of
+        1 -> pure MR.MergeLevel
+        2 -> pure MR.MergeUnion
+        _ -> fail ("[TreeMergeType] Unexpected tag: " <> show tag)
```

Comment:
Related to https://github.com/IntersectMBO/lsm-tree/pull/573#discussion_r1956122054, this might be a clearer way of doing the same thing? Note that this is just a refactoring, not changing the encoding at all, as evidenced by the golden tests.

Now that we have encapsulated the idea into these two instances, we can reconsider whether we think it's a good idea. Maybe the encodings of `LevelMergeType` and `TreeMergeType` should not overlap at all, requiring us to be more explicit about any "reinterpretation" of merge types.